[core-lro] Proposed LRO Engine design

deyaaeldeen opened the pull request:
The PR description just rehashes the [design document](https://github.com/deyaaeldeen/azure-sdk-for-js/blob/lro-design/sdk/core/core-lro/docs/LROEngine.md).

Related to https://github.com/Azure/azure-sdk-for-js/issues/15775

# Modular Support for Long-Running Operations

Long-running operations (LROs) are operations that the service _could_ take a long time to finish processing and they follow a common convention:

- the customer first send an initiation request to the service, which in turn sends back a response, from which the customer can learn how to poll for the status of the operation, if it has not been completed already,
- using their learnings, the customer polls the status of the operation until it is done,
- again, using their learnings, the customer can now get the desired result of the operation once its status says it is done.

Ideally, we can write an algorithm that implements this convention once and use it in all Azure clients for LRO APIs, however, in reality, this convention is implemented differently across Azure services. The good news is that the TypeScript Autorest extension is AFAIK able to generate code that implements those different ones, but this implementation has a couple of limitations:

1. it is located in a few files that the code generator copies into every generated package that has LROs. So if in the future there is a bug fix needed in the LRO logic, the package has to be regenerated with the fix.
2. it works only with clients that use `@azure/core-client`, so clients that use `@azure-rest/core-client` or `@azure/core-http` can not use this implementation as-is.

To fix limitation #1, the most straightforward thing to do is to move those files into `@azure/core-lro`, but without fixing limitation #2 first, `@azure/core-lro` will have to depend on `@azure/core-client` in this case which will force clients that depend on `@azure/core-lro` but not necessarily depend on `@azure/core-client` to transitively depend on the latter, posing concerns about unnecessarily larger bundle sizes.

This document presents a design that fixes limitation #2 and naturally fixes limitation #1 too.

## Things to know before reading

- Some details not related to the high-level concept are not illustrated; the scope of this is limited to the high-level shape and paradigms for the feature area.

## Terminology

- **Azure Async Operation**, **Body**, and **Location** are names for the LRO implementations currently supported in the TypeScript Autorest extension. They vary in how to calculate the path to poll from, the algorithm for detecting whether the operation has finished, and the location to retrieve the desired results from. Currently, these pieces of information can be calculated from the response received after sending the initiation request.

## Why this is needed

The China team is currently waiting for fixing limitation #1 which they regard as a blocker for GAing the TypeScript Autorest extension. Furthermore, having this LRO implementation being part of `@azure/core-lro` and not tied to `@azure/core-client` will significantly help streamline the underway effort to add convenience helpers for LROs in `@azure-rest` clients.

## Proposed design

This document presents a design of an LRO engine to be part of `@azure/core-lro` and could be used by any client regardless of how it is implemented. Furthermore, specific implementations of the engine are also provided to be auto-generated by Autorest.

The design consists of three main pieces:

- an interface, named `LongRunningOperation<T>` which groups various primitives needed to implement LROs
- a class, named `LroEngine`, that implements the LRO engine and its constructor takes as input an object that implements `LongRunningOperation<T>`
- a class that implement `LongRunningOperation<T>` that works with clients that use either `@azure/core-http` and `@azure/core-client`. @joheredi also created one for `@azure-rest/core-client` in https://github.com/Azure/azure-sdk-for-js/pull/15898

### `LongRunningOperation<T>`

This interface contains two methods: **sendInitialRequest** and **sendPollRequest**. I propose to make this interface exported by `@azure/core-lro`.

#### `sendInitialRequest`

This method should be implemented to send the initial request to start the operation and it has the following signature:

```ts
sendInitialRequest: () => Promise<LroResponse<T>>
```

The method does not take the path or the HTTP request method as parameters because they're members of the interface since they're needed to control many aspects of subsequent polling. This is how this method can be implemented:

```ts
public async sendInitialRequest(): Promise<LroResponse<T>> {
  return this.sendOperation(this.args, this.spec); // the class will have sendOperation, args, and spec as private fields
}
```

#### `sendPollRequest`

This method should be implemented to send a polling (GET) request, a request the service should respond to with the current status of the operation, and it has the following signature:

```ts
sendPollRequest: (path: string) => Promise<LroResponse<T>>;
```

This method takes the polling path as input and here is what a simplified implementation would look like:

```ts
  public async sendPollRequest(path: string): Promise<LroResponse<T>> {
    return this.sendOperationFn(this.args, { // the class will have sendOperation, args, and spec as private fields
      ...this.spec,
      path,
      httpMethod: "GET")
    });
  }
```

### `LroEngine`

This class implements the `PollerLike` interface and does the heavy lifting for LROs. I propose to make this class also exported by `@azure/core-lro`. This class has the following type signature:

```ts
class LroEngine<TResult, TState extends PollOperationState<TResult>> extends Poller<TState, TResult>
```

The class also has the following constructor:

```ts
constructor(lro: LongRunningOperation<TResult>, options?: LroEngineOptions);
```

Currently `options` have `intervalInMs` to control the polling interval, `resumeFrom` to enable resuming from a serialized state, and `lroResourceLocationConfig` which could determine where to find the results of the LRO after the operation is finished. Typically, Autorest figures out the value for `LroResourceLocationConfig` from the `x-ms-long-running-operation-options` swagger extension. If there are new arguments to be added to the class, they could be added to the options type.

### [`LroImpl`](https://github.com/deyaaeldeen/autorest.typescript/blob/lro-simplify/src/lroImpl.ts)

This class implements the `LongRunningOperation<T>` interface and I propose to make it auto-generated by Autorest. `LroImpl` needs access to a few pieces: operation specification and operation arguments and a primitive function that can take them as input to send a request and converts the received response into one of type `LroResponse<T>` which has both the flattened and the raw responses.

## Usage examples

### Create an object of `LroImpl`

```ts
const directSendOperation = async (
  args: OperationArguments,
  spec: OperationSpec
): Promise<unknown> => {
  return this.client.sendOperationRequest(args, spec);
};
const sendOperation = async (
  args: OperationArguments,
  spec: OperationSpec
) => {
  let currentRawResponse: FullOperationResponse | undefined = undefined;
  const providedCallback = args.options?.onResponse;
  const callback: RawResponseCallback = (
    rawResponse: FullOperationResponse,
    flatResponse: unknown
  ) => {
    currentRawResponse = rawResponse;
    providedCallback?.(rawResponse, flatResponse);
  };
  const updatedArgs = {
    ...args,
    options: {
      ...args.options,
      onResponse: callback
    }
  };
  const flatResponse = await directSendOperation(updatedArgs, spec);
  return {
    flatResponse,
    rawResponse: {
      statusCode: currentRawResponse!.status,
      body: currentRawResponse!.parsedBody,
      headers: currentRawResponse!.headers.toJSON()
    }
  };
};

const lro = new LroImpl(
  sendOperation,
  { options }, // arguments are just the operation options
  spec
);
```

### Using `LroEngine`

```ts
const pollerEngine = new LroEngine(lro, { intervalInMs: 2000 }); // lro was instantiated in the previous section
const result = pollerEngine.pollUntilDone();
```

## Testing

We have [extensive test suite for LROs](https://github.com/Azure/autorest.typescript/blob/main/test/integration/lro.spec.ts) in the TypeScript code generator repo. I both added those tests here and re-implemented the [lro routes](https://github.com/Azure/autorest.testserver/blob/main/legacy/routes/lros.js) in the Autorest test server. For this to work, I created a [fairly low-level instantiation for `LongRunningOperation<T>` with just `@azure/core-rest-pipeline`](https://github.com/deyaaeldeen/azure-sdk-for-js/blob/lro-design/sdk/core/core-lro/test/utils/coreRestPipelineLro.ts).

xirzec commented on sdk/core/core-lro/docs/LROEngine.md:
```diff
@@ -0,0 +1,213 @@
+
+The design consists of three main pieces:
+
+- an interface, named `LRO<T>` which groups various primitives need to implement LROs
```
Minor thing we should consider - there's been a recent push to not treat acronyms specially in type names, which would make this type `Lro<T>` and similarly change the other instances of `LRO` to `Lro`

Since this name looks a bit odd, I wonder if brevity is working against us and we should actually call the type `LongRunningOperation` for clarity?

deyaaeldeen replied:
I renamed `LRO` to `Lro` in all names.

deyaaeldeen replied:
and renamed `LRO<T>` to `LongRunningOperation<T>`.

xirzec commented on sdk/core/core-lro/review/core-lro.api.md:
```diff
@@ -9,6 +9,76 @@ import { AbortSignalLike } from '@azure/abort-controller';
 export type CancelOnProgress = () => void;
 
+// @public
+export function createGetLROStatusFromResponse<TResult>(lroPrimitives: LRO<TResult>, config: LROConfig, finalStateVia?: FinalStateVia): (rawResponse: RawResponse, flatResponse: TResult) => LROStatus<TResult>;
```
Should we reuse the below type alias for the return type of this function? 

e.g.

```suggestion
export function createGetLROStatusFromResponse<TResult>(lroPrimitives: LRO<TResult>, config: LROConfig, finalStateVia?: FinalStateVia): GetLROStatusFromResponse<TResult>;
```

deyaaeldeen replied:
Done.

xirzec commented on sdk/core/core-lro/review/core-lro.api.md:
```diff
@@ -83,6 +153,21 @@ export interface PollOperationState<TResult> {
 
+// @public
+export type RawHttpHeaders = {
+    [headerName: string]: string;
```
is this collection case sensitive? I wonder if it will cause any trouble, since we did the work to make our headers maps case insensitive

deyaaeldeen replied:
Good question I just call `toJSON()` or `rawHeaders()` from `core-client` and `core-http` respectively to populate this.

deyaaeldeen replied:
I removed `RawHttpHeaders` and just inlined its definition since it is used only once.

xirzec commented on sdk/core/core-lro/review/core-lro.api.md:
```diff
@@ -9,6 +9,76 @@ import { AbortSignalLike } from '@azure/abort-controller';
+// @public
+export interface LROBody extends Record<string, unknown> {
+    properties?: {
+        provisioningState?: string;
```
I assume there is some service-specific reason why provisioningState is in two places on this type?

deyaaeldeen replied:
Yeah I think it is just a case of many LRO conventions out there. However now that I am thinking about it, we could probably hide this and cast to it internally.

deyaaeldeen replied:
I hid `LROBody`.

xirzec commented on sdk/core/core-lro/src/lroPoller/azureAsyncPolling.ts:
```diff
@@ -0,0 +1,71 @@
+          flatResponse,
+          done: false,
+          next: async () => {
+            async function sendFinalRequest(): Promise<LROResponse<TResult> | undefined> {
```
having this function be declared inside `next` feels a little odd, is it possible to pull it out into module scope and pass in the needed args?

deyaaeldeen replied:
Done.

xirzec commented on sdk/core/core-lro/src/lroPoller/bodyPolling.ts:
```diff
@@ -0,0 +1,33 @@
+function getProvisioningState(rawResponse: RawResponse): string {
+  const { properties, provisioningState } = rawResponse.body ?? {};
+  const state: string | undefined = properties?.provisioningState ?? provisioningState;
+  return state?.toLowerCase() ?? "succeeded";
```
we can assume success?

deyaaeldeen replied:
Yeah this is how the autorest test server behaves.

xirzec commented on sdk/core/core-lro/src/lroPoller/bodyPolling.ts:
```diff
@@ -0,0 +1,33 @@
+export function isBodyPollingDone(rawResponse: RawResponse): boolean {
+  const state = getProvisioningState(rawResponse);
+  if (failureStates.includes(state)) {
+    throw new Error(`Provisioning state: ${state}`);
```
can we put some additional details, since I'm not sure it would be obvious to the consumer what to do with an error like this

deyaaeldeen replied:
I just added the prefix: `The long-running operation has failed.`. Let me know if you have another suggestion.

xirzec commented on sdk/core/core-lro/src/lroPoller/stateMachine.ts:
```diff
@@ -0,0 +1,106 @@
+      state.result = flatResponse as TResult;
+      state.isCompleted = true;
+    }
+    return !!state.isCompleted;
```
style nit

```suggestion
    return Boolean(state.isCompleted);
```

deyaaeldeen replied:
Done.

xirzec commented on sdk/core/core-lro/src/lroPoller/stateMachine.ts:
```diff
@@ -0,0 +1,106 @@
+): (pollingURL: string, pollerConfig: PollerConfig) => Promise<LROStatus<TResult>> {
+  return async (path: string, pollerConfig: PollerConfig): Promise<LROStatus<TResult>> => {
+    const response = await lroPrimitives.sendPollRequest(config, path);
+    const retryAfter: string | undefined = response.rawResponse.headers["Retry-After"];
```
this is one of the places where I'm wondering if we can assume all services properly capitalize `Retry-After`

deyaaeldeen replied:
I am not familiar with this tbh, should I also expect `retry-after`? is there any other name the service could use for this?

xirzec replied:
It's possible our services always return `Retry-After` and this will never be a problem. I think it's just one of those potential issues, where it is perfectly valid for the server to return `retry-after` or `rEtRy-AfTeR` and it would still be valid.

If you're using toJson() of our HttpHeaders classes then I think we normalize to all lowercase

deyaaeldeen replied:
I see, this is a bug then. I will fix it.

deyaaeldeen replied:
used `retry-after` instead and added a comment on the `headers` field that it is already normalized.

xirzec commented on sdk/core/core-lro/src/lroPoller/stateMachine.ts:
```diff
@@ -0,0 +1,106 @@
+        ? calculatePollingIntervalFromDate(new Date(retryAfter), pollerConfig.intervalInMs)
+        : retryAfterInMs;
+    }
+    return response!;
```
Why do we need the `!` here? I thought `sendPollRequest` always returns a value

deyaaeldeen replied:
You're right. It fell through the cracks when I was doing the decoupling from core-client. Before, it was being set inside the callback.

deyaaeldeen replied:
Removed `!`.

xirzec commented on sdk/core/core-lro/src/lroPoller/requestUtils.ts:
```diff
@@ -0,0 +1,61 @@
+  return rawResponse.headers["operation-location"];
+}
+
+function getAzureAsyncoperation(rawResponse: RawResponse): string | undefined {
```
nit

```suggestion
function getAzureAsyncOperation(rawResponse: RawResponse): string | undefined {
```

deyaaeldeen replied:
Done.

xirzec commented on sdk/core/core-lro/src/lroPoller/requestUtils.ts:
```diff
@@ -0,0 +1,61 @@
+ * where both azure-asyncoperation and location could be present in the same response but
+ * azure-asyncoperation should be the one to use for polling.
+ */
+export function getPollingURL(rawResponse: RawResponse, defaultPath: string): string {
```
nit
```suggestion
export function getPollingUrl(rawResponse: RawResponse, defaultPath: string): string {
```

deyaaeldeen replied:
Done.

xirzec commented on sdk/core/core-lro/src/lroPoller/operation.ts:
```diff
@@ -0,0 +1,96 @@
+    if (!state.isCompleted) {
+      if (this.GetLROStatusFromResponse === undefined) {
+        if (state.config === undefined) {
+          throw new Error("Bad state: LRO mode is undefined");
```
If this gets surfaced to the user, we should perhaps provide some guidance on how to handle it?

xirzec commented on sdk/core/core-lro/src/lroPoller/operation.ts:
```diff
@@ -0,0 +1,96 @@
+        this.GetLROStatusFromResponse = createPollForLROStatus(this.lro, state.config);
+      }
+      if (state.pollingURL === undefined) {
+        throw new Error("Bad state: polling URL is undefined");
```
Same thought here

deyaaeldeen replied:
I added `Please check if the serialized state is well-formed`.

xirzec commented on sdk/core/core-lro/src/lroPoller/operation.ts:
```diff
@@ -0,0 +1,96 @@
+      }
+      const currentState = await this.GetLROStatusFromResponse(
+        state.pollingURL,
+        this.pollerConfig!
```
why can we assume `pollerConfig` has been assigned here?

deyaaeldeen replied:
Because it has been set inside the constructor. The operation polling should not start until the customer calls `poll` or its sisters AFAIK.

xirzec commented on sdk/core/core-lro/src/lroPoller/operation.ts:
```diff
@@ -0,0 +1,96 @@
+        state.pollingURL = getPollingURL(currentState.rawResponse, state.pollingURL);
+      }
+    }
+    if (options?.fireProgress !== undefined) {
+      options.fireProgress(state);
+    }
```
I think this works

```suggestion
    options?.fireProgress?.(state);
```

deyaaeldeen replied:
Done.

xirzec commented on sdk/core/core-lro/src/lroPoller/models.ts:
```diff
@@ -0,0 +1,162 @@
+};
+
+export const successStates = ["succeeded"];
+export const failureStates = ["failed", "canceled", "cancelled"];
```
I'm guessing services aren't consistent with how they spell `cancelled`? ☹️

deyaaeldeen replied:
Yup

xirzec commented on sdk/core/core-lro/review/core-lro.api.md:
```diff
@@ -83,6 +153,21 @@ export interface PollOperationState<TResult> {
+}
+
+// @public
+export const terminalStates: string[];
```
out of curiosity, why do we need to expose this?

deyaaeldeen replied:
They're being used by `coreClientLRO` and `coreHttpLRO`. Maybe we can get by with hiding it and duplicating there?

deyaaeldeen replied:
I hide it.

xirzec commented on sdk/core/core-lro/src/lroPoller/locationPolling.ts:
```diff
@@ -0,0 +1,23 @@
+function isLocationPollingDone(rawResponse: RawResponse): boolean {
+  const code = rawResponse.statusCode;
+  if (![202, 200].includes(code)) {
+    throw new Error(`Operation failed`);
```
If this bubbles up, we should think about how to message it better, like `Received unexpected HTTP status code ${code} while polling. This may indicate a server issue.`

We should also figure out how we recommend devs recover from such a state, same with other exceptions we through in these internals

deyaaeldeen replied:
Done. Not sure about the how to recover from it part though, are you saying we should customize the message based on the http code?

xirzec replied:
For the first cut I don't think we need to go too crazy. To start, just think about what you'd want to see in a GitHub issue when a consumer is hitting a bug here because the server is misbehaving.

deyaaeldeen replied:
Done.

xirzec commented on sdk/core/core-lro/docs/LROEngine.md:
```diff
@@ -0,0 +1,213 @@
+
+### `coreClientLRO` and `coreHttpLRO`
+
+I propose to make these classes auto-generated by Autorest. These classes implement `LRO<T>` and various bits of `coreHttpLRO` have been showed earlier. These classes will need access to a few pieces: operation specification and operation arguments and a primitive function that can take them as input to send a request and also converts the response into one of type `LROResponse<T>` which has both ...
```
out of curiosity, roughly how big are these classes?

deyaaeldeen replied:
You can find those classes in https://github.com/Azure/autorest.typescript/pull/1043/. They're the first two files there.

xirzec replied:
Hmm not too bad (~300 lines). I wonder if after core-http is deprecated we should revisit pulling in `coreClientLRO`

deyaaeldeen replied:
I would say we do not want to pull in `coreClientLRO` because we do not want to make `core-lro` depend on `core-client`, since [`@azure-rest/core-lro-helper`](https://github.com/Azure/azure-sdk-for-js/pull/15898) is going to be a thing. But perhaps we could create a `@azure/core-lro-helper` or something like that.

xirzec commented on sdk/core/core-lro/docs/LROEngine.md:
```diff
@@ -0,0 +1,213 @@
+  args: coreClient.OperationArguments,
+  spec: coreClient.OperationSpec
+): Promise<unknown> => {
+  return this.client.sendOperationRequest(args, spec);
```
where does `this` come from?

deyaaeldeen replied:
it refers to the object that has this operation, this is a rough code snippet copied from [here](https://github.com/Azure/autorest.typescript/blob/5f7acf24417c9c4b8bdac7f59726792d90c537f6/test/integration/generated/lro/src/operations/lROs.ts#L194).

deyaaeldeen replied:
Now that I think about it, we should probably pull that into a function or something instead of inlining it in every operation like this.

deyaaeldeen replied:
I opened https://github.com/Azure/autorest.typescript/issues/1079.

xirzec reviewed (commented):
Great work on this!

Overall, I think this is a good strategy, though I am curious how many packages are using LROs and `core-http` still - will some of this abstraction cause unnecessary complexity once we're migrated fully off of core-http?

Most of my comments are clarifying or style nits.

deyaaeldeen commented:
@xirzec 

> though I am curious how many packages are using LROs and core-http still

AFAIK, none. `core-http` is used in Track 2 packages and we write LROs there by hand. Hopefully the recent improvements to auto-generated LROs will change this.

> will some of this abstraction cause unnecessary complexity once we're migrated fully off of core-http?

We already see the abstraction in action for core-client and core-http in https://github.com/Azure/autorest.typescript/pull/1043 and I do not think there is any complexity due to core-http or its phasing out. In fact, there is slight complexity because of the dancing around the callback mechanism in core-client instead, (e.g. the need for the `initializeState` param to `sendInitialRequest` that returns whether the LRO has finished so that the customer-provided callback can be called at this point).

deyaaeldeen commented on sdk/core/core-lro/docs/LROEngine.md:
```diff
@@ -0,0 +1,213 @@
+This method is only needed to potentially retrieve the provisioned azure resource in the _Azure Async Operation_ mode after the operation has finished and it should be different from the `sendPollRequest` one in two things:
+
+- should set the obscure `shouldDeserialize` option needed for doing custom de-serialization on the result if it is supported by the client
+- we already know the operation has finished at this point, so `done` can be set to `true` right away without calling `getLroStatusFromResponse`
```
I think we could it make it easier for clients to implement by setting done to true internally, and change the return type to be just `LroResponse` instead of `LroStatus`.

sadasant commented on sdk/core/core-lro/review/core-lro.api.md:
```diff
@@ -9,6 +9,67 @@ import { AbortSignalLike } from '@azure/abort-controller';
+
+// @public
+export interface LroResponse<T> {
+    // (undocumented)
```
Hi hi! Please make sure nothing is left undocumented in this file

deyaaeldeen replied:
Done.

sadasant commented on sdk/core/core-lro/test/utils/router/paramRoutes.ts:
```diff
@@ -0,0 +1,384 @@
+      (initialCode === 200 &&
+        initialState === "Accepted" &&
+        finalState === "Canceled" &&
+        finalCode === 200)
```
Are we sure these strings are going to be properly cased every time? I wonder if it would be better to always lowercase them.

sadasant commented on sdk/core/core-lro/test/utils/router/paramRoutes.ts:
```diff
@@ -0,0 +1,384 @@
+  const pieces = path.substr(1).split("/");
+  if (pieces.length === 5) {
+    try {
+      const initialCode: number = JSON.parse(pieces[1]);
```
Can we wrap all of the JSON.parse calls into try catch blocks ? (or to have a parser function that would deal with the try-catch of all of the JSON.parses). Ideally we would have known failure cases with known error messages in the cases the JSONs are misshapen.

sadasant commented on sdk/core/core-lro/src/lroEngine/lroEngine.ts:
```diff
@@ -0,0 +1,37 @@
+  constructor(lro: LongRunningOperation<TResult>, options?: LroEngineOptions) {
+    const { intervalInMs = 2000, resumeFrom } = options || {};
+    const state: TState & ResumablePollOperationState<TResult> = resumeFrom
+      ? JSON.parse(resumeFrom).state
```
Can we wrap all of the JSON.parse calls in try-catch blocks? Or to have a utility function that does so in a single place? Ideally we would have a standard error message or a set of messages for the known cases of misshapen JSON values.

deyaaeldeen replied:
This is the only place where  `JSON.parse` is called. I added trying logic.

sadasant commented on sdk/core/core-lro/package.json:
```diff
@@ -102,13 +99,10 @@
   },
   "devDependencies": {
     "@azure/core-http": "^2.0.0",
```
Can we remove core-http from the devDependencies?

deyaaeldeen replied:
This is out of scope for this PR but I agree we should remove it. However doing so will require rewriting test files. I can do that in another PR though.

sadasant commented on sdk/core/core-lro/docs/LROEngine.md:
```diff
@@ -0,0 +1,214 @@
+
+Long-running operations (LROs) are operations that the service _could_ take a long time to finish processing and they follow a common convention:
+
+- the customer first send a an initiation request to the service, which in turn sends back a response, from which the customer can learn how to poll for the status of the operation, if it has not been completed already,
```
Please pass this text through Word or Grammarly to ensure it’s properly cased, and to ensure it’s free from typos

deyaaeldeen replied:
Done.

sadasant commented on sdk/core/core-lro/docs/LROEngine.md:
```diff
@@ -0,0 +1,214 @@
+
+In order to fix limitation #1, the most straightforward thing to do is to move those files into `@azure/core-lro`, but without fixing limitation #2 first, `@azure/core-lro` will have to depend on `@azure/core-client` in this case which will force clients that depend on `@azure/core-lro` but not necessarily depend on `@azure/core-client` to transitively depend on the latter, posing concerns about  ...
+
+This document presents a design that fixes limitation #2 and naturally fixes limitation #1 too.
```
While I appreciate the context, I’m not sure if this design proposal should be in the repository at the end. I would expect the design proposal to be somewhere else, and the docs in the core-lro repo to be about documentation.

deyaaeldeen replied:
We did that before for core v2: https://github.com/Azure/azure-sdk-for-js/blob/main/sdk/core/core-rest-pipeline/documentation/core2.md. However, I am open to move/remove it if there is a consensus on this.

sadasant replied:
oh ok ok! I’m ok leaving this here then. 👍

sadasant commented on sdk/core/core-lro/docs/LROEngine.md:
````diff
@@ -0,0 +1,214 @@
+const pollerEngine = new LroEngine(lro, { intervalInMs: 2000 }); // lro was instantiated in the previous section
+const result = pollerEngine.pollUntilDone();
+```
+
````
Wouldn’t it be possible to create LROs specific to the known Azure specs with the core-lro classes that are already available?

sadasant commented on sdk/core/core-lro/review/core-lro.api.md:
```diff
@@ -9,6 +9,65 @@ import { AbortSignalLike } from '@azure/abort-controller';
+export function createGetLroStatusFromResponse<TResult>(lroPrimitives: LongRunningOperation<TResult>, config: LroConfig, lroResourceLocationConfig?: LroResourceLocationConfig): GetLroStatusFromResponse<TResult>;
+
+// @public
+export type GetLroStatusFromResponse<T> = (rawResponse: RawResponse, flatResponse: T) => LroStatus<T>;
```
Can we use more specific `T` types? like how we were doing with `TResult`, for example?

deyaaeldeen replied:
no longer part of the public API

sadasant commented on sdk/core/core-lro/review/core-lro.api.md:
```diff
@@ -9,6 +9,65 @@ import { AbortSignalLike } from '@azure/abort-controller';
+export type LroMode = "AzureAsync" | "Location" | "Body";
+
+// @public
+export type LroResourceLocationConfig = "azure-async-operation" | "location" | "original-uri";
```
Would extensible enums work here? (also on LroMode)?

deyaaeldeen replied:
I do not think so because this is a defined swagger option. `LroMode` is no longer part of the public surface.

sadasant commented on sdk/core/core-lro/src/lroEngine/azureAsyncPolling.ts:
```diff
@@ -0,0 +1,82 @@
+  return successStates.includes(state);
+}
+
+async function sendFinalRequest<TResult>(
```
Please document these functions! Someone new learning this could have a hard time understanding the purpose of these.

sadasant replied:
Please document all the internal functions that aren’t self-evident, if that makes sense

deyaaeldeen replied:
Done.

sadasant commented on sdk/core/core-lro/src/lroEngine/operation.ts:
```diff
@@ -0,0 +1,103 @@
+   *      - Strategies will always use the latest URI for polling if provided otherwise
+   *        the last known one
+   */
+  async update(options?: {
```
This is really the gem of all things LRO. We should either add a logger here that would showcase the behavior on debug, or add a guide on how to debug LROs that would point to how to add log lines and debugger instructions on this specific function.

deyaaeldeen replied:
@xirzec I do not know much about logging, do you have ideas about whether I should do something special here to help have a good debugging experience?

xirzec replied:
Perhaps to start, you can import `createClientLogger` from `@azure/logger` to make a logger for `core-lro`, then call `verbose` on it to log whats going on

deyaaeldeen replied:
Done.

sadasant commented on sdk/core/core-lro/src/lroEngine/operation.ts:
```diff
@@ -0,0 +1,103 @@
+      if (this.getLROStatusFromResponse === undefined) {
+        if (state.config === undefined) {
+          throw new Error(
+            "Bad state: LRO mode is undefined. Please check if the serialized state is well-formed."
```
do you mind if we move the error messages outside of this function into a common place? could be right above the function. I think it will make it easier to keep these messages on sync over time.

deyaaeldeen replied:
I did some edits there but felt like pulling out stuff will not help make this more readable.

sadasant commented on sdk/core/core-lro/test/engine.spec.ts:
```diff
@@ -0,0 +1,538 @@
+      const result = await runMockedLro("PUT", "/put/201/creating/succeeded/200");
+      assert.deepEqual(result.properties?.provisioningState, "Succeeded");
+      assert.deepEqual(result.id, "100");
+      assert.deepEqual(result.name, "foo");
```
Beautiful tests!

sadasant commented on sdk/core/core-lro/test/utils/router.ts:
```diff
@@ -0,0 +1,98 @@
+    poller.onProgress(onProgress);
+  }
+  return poller.pollUntilDone();
+}
```
This file really showcases what the code generator will be generating. I think here’s the beauty of the design you’ve come up with. We should help developers find this file somehow, perhaps in the guide on how to use these new classes.

sadasant reviewed (approved):
I really like this design. I’ve left some comments, but otherwise this looks amazing! Thank you for the hard work.

sadasant commented:
Please don’t merge this until @xirzec , @chradek or @bterlson have approved this.

deyaaeldeen commented:
@bterlson and @xirzec I simplified the design a lot in https://github.com/Azure/azure-sdk-for-js/pull/15949/commits/3fb912068c505d6657d4a94aa0d3f14ed3d31d29. This time the client code has to implement two methods only instead of three: `sendInitialRequest` and `sendPollRequest` and I ditched `retrieveAzureAsyncResource` by replacing it with `sendPollRequest`. Furthermore, `sendPollRequest` was simplified to take the request path and an `isDone` predicate as input. The `isDone` predicate is useful for `@azure/core-client` clients as explained in the design document. I removed the `LroMode` parameter after I realized I can build a `sendPollRequest` without it and pass all the tests. You can see the design in action in code gen here: https://github.com/Azure/autorest.typescript/pull/1099.

This change dramatically decreases the public surface because many types and helper functions are no longer needed: https://github.com/Azure/azure-sdk-for-js/pull/15949/commits/3fb912068c505d6657d4a94aa0d3f14ed3d31d29#diff-6f92516f198f704d6b14b9017fa3f9acdb4d30dc2753cc1fcb51eee9704f30dc. Please give it another look.

deyaaeldeen commented on sdk/core/core-lro/test/utils/coreRestPipelineLro.ts:
```diff
@@ -0,0 +1,31 @@
+    return response;
+  }
+
+  public async sendPollRequest(url: string): Promise<LroResponse<T>> {
```
The client can ignore calling the `isDone` predicate.

deyaaeldeen commented on sdk/core/core-lro/src/lroEngine/operation.ts:
```diff
@@ -0,0 +1,130 @@
+  }
+}
+
+function memoize<TInput extends unknown[], TResult>(
```
perhaps this should be moved into a utility module.

xirzec commented on sdk/core/core-lro/src/lroEngine/azureAsyncPolling.ts:
```diff
@@ -0,0 +1,77 @@
+    case "original-uri":
+      return lro.sendPollRequest(lro.requestPath, () => true);
+    case "azure-async-operation":
+      return Promise.resolve(undefined);
```
this is an async function, so you shouldn't need to wrap `undefined` inside `Promise.resolve`

deyaaeldeen replied:
Done.

xirzec commented on sdk/core/core-lro/src/lroEngine/lroEngine.ts:
```diff
@@ -0,0 +1,46 @@
+
+  constructor(lro: LongRunningOperation<TResult>, options?: LroEngineOptions) {
+    const { intervalInMs = 2000, resumeFrom } = options || {};
+    function deserializeState(
```
having a function declared inside the constructor body feels a bit strange. Can we hoist this out of the class into module scope?

deyaaeldeen replied:
Done.

xirzec commented on sdk/core/core-lro/src/lroEngine/lroEngine.ts:
```diff
@@ -0,0 +1,46 @@
+    super(operation);
+
+    this.intervalInMs = intervalInMs;
+    operation.setPollerConfig(this as any);
```
this looks a little fishy

xirzec replied:
this `intervalInMs` is a private, could this be `operation.setPollerConfig({intervalInMs: this.intervalInMs});` ?

deyaaeldeen replied:
I agree it is fishy but I needed to do this because I wanted to be able to set the `intervalInMs` member of `LroEngine` inside the operation using the `retry-after` header value. I am open to ideas to make this cleaner.

xirzec replied:
perhaps you could have a private object, e.g.

```
this.pollerConfig = {
   intervalInMs
}
```

then pass that object along, and since you still have a reference to it, you can update it later.

deyaaeldeen replied:
For some reason, I thought `intervalInMs` is a public member so that was why I did it that way. I fixed it now.

xirzec commented on sdk/core/core-lro/src/lroEngine/lroEngine.ts:
```diff
@@ -0,0 +1,46 @@
+    }
+    const state: TState & ResumablePollOperationState<TResult> = resumeFrom
+      ? deserializeState(resumeFrom)
+      : ({} as any);
```
?
```suggestion
      : ({} as TState & ResumablePollOperationState<TResult>);
```

deyaaeldeen replied:
Done.

xirzec commented on sdk/core/core-lro/src/lroEngine/operation.ts:
```diff
@@ -0,0 +1,130 @@
+    }
+
+    if (!state.isCompleted) {
+      if (this.poll === undefined || this.getLroStatusFromResponse === undefined) {
```
style nit: it seems like most of these explicit undefined checks are unnecessary, since there are no valid falsy values. 

I find this easier to read:

```suggestion
      if (!this.poll || !this.getLroStatusFromResponse) {
```

deyaaeldeen replied:
Done.

xirzec commented on sdk/core/core-lro/src/lroEngine/operation.ts:
```diff
@@ -0,0 +1,130 @@
+function memoize<TInput extends unknown[], TResult>(
+  f: (...args: TInput) => TResult
+): () => (...args: TInput) => TResult {
+  return (): ((...args: TInput) => TResult) => {
+    let result: TResult | undefined = undefined;
+    return (...args: TInput): TResult => {
+      if (result === undefined) {
+        result = f(...args);
+      }
+      return result;
+    };
+  };
```
I feel like there is an extra level in here that I don't understand. 

First you pass it a function to memoize, then you call it again to get the actual memoized function? It also doesn't seem to be checking the args at all, it just caches the last result regardless of if the args change?

I would expect it to work like this:

```ts
function add(a: number, b:number): number {
   return a + b;
}
const memoizedAdd = memoize(add);
memoizedAdd(1, 2); // returns 3 after calling add
memoizedAdd(1, 2); // returns 3 without calling add
memoizedAdd(3, 4); // returns 7 after calling add
```

deyaaeldeen replied:
I just did not go all the way on memoization for each input because I did not need it but it is simple enough I could as well just add it. Also, you're right, I accidentally introduced an extra outer function.

deyaaeldeen replied:
Done.

xirzec commented on sdk/core/core-lro/src/lroEngine/requestUtils.ts:
```diff
@@ -0,0 +1,94 @@
+  return {};
+}
+
+export class RestError extends Error {
```
Since this `RestError` has differences from `RestError` in either `core-http` or `core-rest-pipeline`, we should call it something else to avoid confusion.

deyaaeldeen replied:
Renamed it to `SimpleRestError` and unexported it.

xirzec commented on sdk/core/core-lro/docs/LROEngine.md:
````diff
@@ -0,0 +1,167 @@
+sendPollRequest: (path: string, isDone: (response: LroResponse<T>) => boolean) => Promise<LroResponse<T>>;
+```
+
+This method takes two things as input, the request path as you would expect but also a predicate `isDone` that can tell you whether the operation has finished. Similar to the `initializeState` in `sendInitialRequest`, this predicate is needed in clients that use `@azure/core-client` because the customer-provided callback will have to be called if the predicate returns true. `isDone` is implemente ...
````
I'm a bit confused by the `isDone` param, since I can't seem to find meaningful usage of it anywhere in this PR. What scenarios is it needed for and can we add some tests for these?

deyaaeldeen replied:
It is used in the auto-generated `coreClientLro.ts`: https://github.com/deyaaeldeen/autorest.typescript/blob/8e1ab5b7575bb0aaaf1d02ee93a0be0eff04807f/src/coreClientLro.ts#L43. This works by the fact that the integration tests in code gen pass since all assertion logic is done inside those callbacks. Here is the PR: https://github.com/Azure/autorest.typescript/pull/1099

xirzec replied:
Why do we only call `onResponse` when it's done? The point of `onResponse` is to debug *every* HTTP call, not the last one.

deyaaeldeen replied:
Ok I removed the `isDone` predicate param.

xirzec reviewed (changes requested):
I'm very excited about this PR! There are just a couple things that I would like to get some clarity on before signing off.

xirzec commented on sdk/core/core-lro/src/lroEngine/operation.ts:
```diff
@@ -0,0 +1,128 @@
+function memoize<TInput extends unknown[], TResult>(
+  f: (...args: TInput) => TResult
+): (...args: TInput) => TResult {
+  const result: Map<TInput, TResult> = new Map();
```
how long will this map hold onto response objects? I could see this getting costly. Furthermore, `has` isn't going to work well with arrays, since each array of args will have a unique identity, even if each element is the same. 

I think instead of trying to write a generic `memoize` helper, it would be simpler to do specific work to memoize the `getLroStatusFromResponse` function directly.

deyaaeldeen replied:
I removed `isDone` so this is no longer needed.

xirzec commented on sdk/core/core-lro/src/lroEngine/azureAsyncPolling.ts:
```diff
@@ -0,0 +1,80 @@
+  if (isUnexpectedPollingResponse(rawResponse) || failureStates.includes(state)) {
+    throw new Error(`The long running operation has failed. The provisioning state: ${state}.`);
+  }
+  return successStates.includes(state);
```
I'm a little curious about why failure is exceptional instead of being viewed as a final state, like success. Is failure never expected to happen normally, so this method is really just disambiguating between success and in progress?

deyaaeldeen replied:
The code gen crew agreed on throwing if the operation reached a failure state and this function just implements this expectation. I think it is mostly because other languages started this way so JS has to be consistent with them. To me, I think this is not the best thing to do because all the customer gets is an exception message which I think it makes rehydration harder.

xirzec reviewed (approved):
I feel great about the new design! Nice work.

sarangan12 commented on sdk/core/core-lro/docs/LROEngine.md:
```diff
@@ -0,0 +1,158 @@
+
+## Testing
+
+We have [extensive test suite for LROs](https://github.com/Azure/autorest.typescript/blob/main/test/integration/lro.spec.ts) in the TypeScript code generator repo. I both added those tests here and re-implemented the [lro routes](https://github.com/Azure/autorest.testserver/blob/main/legacy/routes/lros.js) in the Autorest test server. For this to work, I created a [fairly low-level instantiation  ...
```
@deyaaeldeen This link is broken. This causes release CI to be broken. Could you fix them? 

Also 2 links in this page is pointing to your fork instead of Azure fork. Is that expected?
